### PR TITLE
Send reply to group/channel

### PR DESCRIPTION
**Pull request by ozbek**
Currently, bot always sends replies to the user by PM regardless of the origin of the message (PM or group). This enables bot to send replies to group.

### CODE REVIEW

**Comment by ozbek** [comment]
@rougeth, any thoughts on this PR? Maybe the code is not up-to min requirements? Or the concept is wrong?

**Comment by rougeth** [comment]
Hi @ozbek, sorry about the delay. Thanks for the PR, I new this issue but didn't have time to think about it.

My concern here is that `bottery.message.Message` should be generic for all platforms. I don't think `chat` attr make sense on FB Messenger.

What's your thought about this issue?

**Comment by ozbek** [comment]
Thanks. And you are right, this indeed will fail for other other platforms. Can't come up with something useful right away...

**Comment by nicoddemus** [comment]
Throwing an idea for discussion:

Perhaps `bottery.message.Message` should be an interface instead, and the actual implementation should be created/returned by the platform/backend. This way each platform/backend will have its own implementation being free to add features specific to that platform: Telegram would return a `TelegramMessage` having a `chat` attribute for example.

Chatbots that target a certain platform then may decide to use those new features.

**Comment by leportella** [comment]
@nicoddemus this seams to be a good approach to follow, but will it affect the patterns handlers?

**Comment by rougeth** [comment]
@leportella I cannot see it affecting the patterns handlers because all `[platform]Message` classes will extend from the base interface, right?

**Comment by nicoddemus** [comment]
> patterns handlers because all [platform]Message classes will extend from the base interface

Yep that's what I was thinking. Pattern handles would match based on the attributes of the interface.

**Comment by leportella** [comment]
Seems a good idea. @ozbek would you like to do so?

**Comment by ozbek** [comment]
Err... I am afraid I am not proficient enough to go that far. I'm happy if someone else takes this and ready to close this PR.

File: bottery/platform/telegram/engine.py [hunk 1]
```diff
@@ -127,7 +149,10 @@ def build_message(self, data):
         # TODO: Test if the view returned something or not
         response = await self.get_response(view, message)
 
+        chat_id = message.user.id
+        if not message.chat.type == 'private':
+            chat_id = message.chat.id
```
Comment (rougeth) [inline bottery/platform/telegram/engine.py]: Can you move this `if` to a function and then create a unit test for it? It makes easier to tests and refactor and `message_handler` continues to be a small function with limited responsibilities.

**Comment by rougeth** [comment]
@ozbek no worries :) I made a comment about the `if` that checks the id. Once that's ready, we can merge this PR and work on the message interface later.

**Comment by leportella** [comment]
You will need to make a rebase as well :)

you can:

`git checkout master`
`git pull`
`git checkout your-branch`
`git rebase master`

:D

File: bottery/platform/telegram/engine.py (hunk 1 above)
Comment (nicoddemus) [reply bottery/platform/telegram/engine.py]: > ... and then create a unit test for it?

😍

**Comment by leportella** [comment]
You should do a rebase on your branch as well :)

You can:

`git remote add upstream git@github.com:rougeth/bottery.git`
`git fetch upstream`
`git rebase upstream/master`

This way you will end this message *This branch is out-of--date...*

:D Hope it helps

**Comment by ozbek** [comment]
Thanks, let me try.

**Comment by rougeth** [comment]
@ozbek took the liberty and just fixed all the conflicts :)

**Comment by ozbek** [comment]
All got messed up after rebase. Thanks :)

So far so good? Now need to write the tests?

**Comment by rougeth** [comment]
@ozbek yep, also, take a look at the build at Travis, two lines with problems when running flake8 :)
https://travis-ci.org/rougeth/bottery/jobs/325616042

**Comment by rougeth** [comment]
Hi @ozbek! Needing any help? I saw the tests you made are actually testing other functions of Bottery, not the ones you created.

**Comment by ozbek** [comment]
Hi. Yes, in fact, I do need help. I am not sure if I am on the right track.

File: tests/test_platform_telegram.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    api = TelegramAPI('token', aiohttp.ClientSession)
+    with pytest.raises(AttributeError):
+        api.non_existent_method()
+
+
+@pytest.mark.asyncio
+async def test_platform_telegram_api_get_updates():
+    api = TelegramAPI('token', aiohttp.ClientSession)
+    with pytest.raises(TypeError):
+        await api.get_updates()
```
Comment (rougeth) [inline tests/test_platform_telegram.py]: This one here should be a bit different. It's actually making a request to Telegram API. Would be better mock session and check if it was called.

File: tests/test_platform_telegram.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    chat = type('TelegramChat', (), {'type': 'private'})
+    message = type('Message', (), {'user': user, 'chat': chat})
+    engine = TelegramEngine
+
+    assert engine.get_chat_id(engine, message) == '123'
+
+
+def test_platform_telegram_engine_get_chat_id_group():
+    chat = type('TelegramChat', (), {'id': '456', 'type': 'group'})
+    message = type('Message', (), {'chat': chat})
```
Comment (rougeth) [inline tests/test_platform_telegram.py]: Both test here are quite similar. I believe using a fixture to the Message instance and pytest parametrize feature would make it better. Something like (considering we have engine and message fixtures):
```
# this should work but I didn't test
@pytest.mark.parametrize('chat_type,id_expected', [
    ('group', 456),
    ('private', 123),
])
def test_platform_telegram_engine_get_chat_id(chat_type, id_expected, engine, message):
    setattr(message.chat, 'type', chat_type)
    assert engine.get_chat_id(message) == id_expected
```
What do you think?

**Comment by ozbek** [comment]
@rougeth, can you help with the travis build error please?

`ERROR: InvocationError: '/home/travis/build/rougeth/bottery/.tox/check/bin/isort bottery tests --recursive --check-only --diff'`

**Comment by rougeth** [comment]
Sure! `isort` checks the order the imports in all .py files. In this case, the ouput is this one:
```
 from bottery.message import Message
-from bottery.platform.telegram.engine import TelegramChat
-from bottery.platform.telegram.engine import TelegramEngine
-from bottery.platform.telegram.engine import TelegramUser
+from bottery.platform.telegram.engine import (TelegramChat, TelegramEngine,
+                                              TelegramUser)
```
The last approach should be used instead of the first one.

**Comment by ozbek** [comment]
Thanks, that seems to fix the travis build...

The branch is behind again. Do you want me to rebase it again?

**Comment by rougeth** [comment]
@ozbek I'm merging your PR now. We will need to make some changes like we discussed here and I will open an issue for that.
Thank you very much!

**Pull request merged**